### DISCUSSION TITLE
Filter-extensions: how to remove an entry of the `format-links`-element for an HTML-page?

### DISCUSSION THREAD

**Discussion by Gewerd-Strauss** [quarto-dev/quarto-cli#10375] [Q&A]
### Description

Hello,

I am currently working on a small extension that does the following:

1. Check the yaml-key `suppress-formats`. 
2. For any format listed therein, if it is rendered to via `quarto render file --to <format>`, its body will be stripped. 

This will then leave a document consisting only of its frontmatter and whatever that generates, but lacking its document body itself:

HTML: 
![grafik](https://github.com/user-attachments/assets/7fb540e2-1d2b-4ef4-a531-098f3055c799)

PDF: 
![grafik](https://github.com/user-attachments/assets/f631bed0-7729-4139-b106-24d00888940c)

DOCX: 
![grafik](https://github.com/user-attachments/assets/be98d6f9-9cc9-4eb9-9b50-e6187012fe8d)

The current code, as well as testing-documents can be found at https://github.com/Gewerd-Strauss/suppress-format-render

Take the following input doc: 

```qmd
---
title: "basic example"
subtitle: "Suppresses pdf, renders html and docx"
format-links: true
suppress-formats:
  - pdf
format: 
  html: default
  pdf: default
  docx: default
---



## Heading

This filter suppresses rendering of the main body of a document when the YAML-key contains the name of the currently executed format.

Note: Format Aliases can be used, as documented in the [quarto-lua-api-documentation](https://quarto.org/docs/extensions/lua-api.html#format-detection)

```

This all works.

For HTML-output, I now want to remove the `other formats`-entry for any format which was stripped this way. 
However, I cannot figure out how to do so. I can't seem to access the data via `doc.meta.format_links`/`doc.meta["format-links"]`, or really at all, seemingly.

I have not yet decided how to handle the scenario where `format-links` is not a boolean `T`/`F`, but a list of linked-to formats such as:

```qmd
---
title: "basic example"
subtitle: "Suppresses pdf, renders html and docx"
format-links: [pdf]
suppress-formats:
  - pdf
format: 
  html: default
  pdf: default
  docx: default
---
```

But as I mostly need the use-case of `format-links: true` in my current scenario, I am not considering this yet.

Is there any way I can modify this element of the HTML-output?

Thank you.
Sincerely,
~Gw

**Comment by mcanouil** [comment]
`format-links` is handle by Quarto internals (TypeScript) and is not accessible to Lua.

Note that `format-links` can be set for each format and not necessarily at the global level.

**Reply by Gewerd-Strauss** [reply]
> Note that format-links can be set [...] not necessarily at the global level.

That is interesting. 

So I could write a pre-render script to modify my input docs to dynamically populate `format-links`? Could I modify the file via the filter itself?




Multiple questions related to this:
1. Am I correct that I can toggle it on a per-directory- or per-document basis via frontmatter?
2. As I understand https://quarto.org/docs/extensions/filters.html#overview, I cannot modify the state of the frontmatter via the filter itself, correct? When the filter is running, rendering has commenced already, and I would need to do so prior to that point?

As I understand it, the following process should work:


1. Determine which formats are being rendered to, store that info.
2. Iterate over all input-docs; and check the respective frontmatter.
3. On each document, populate `format-links` with the formats which are being rendered to which are not present in `suppress-formats`.
4. Begin rendering.

---

Implementation:

I know I can do this in a pre-render script, because I already transform my raw markdown input-files of format `\_{2,}(?<name>\w*)\.md` into qmd-files with stripped underscores: `%name%.qmd`. So I should just as well be able to insert an additional intermediary step modifying the frontmatter.

But that would require a pre-render script for implementation, which would defeat the point of this being a filter-extension exclusively.

It would be better if I could perform this step by a lua-filter on the files themselves. But again, I suspect lua is not capable of accessing that information "early enough", correct? 
Would executing the filter before the "quarto"-filters allow modification "early enough"?

```qmd
---
filters:
  - path/to/suppress-format-render.lua
  - quarto
---

A doc.
```


---

And a final question:
I don't think I can suppress code-execution per-format to save time, correct? I.e. embedded code getting executed for each format can't be selectively enabled based on frontmatter-state,right? I mean I could set `eval: false` dynamically by modifying an early code-chunk and setting the option, but that would obviously break if I have inline R-code referencing generated data, which should then error out?


Thus, the most I should be able to save this way is the overhead of the document itself being run through pandoc's writer, under the assumption that an empty doc gets processed faster than a populated one.
